staging/src/k8s.io/client-go/util/workqueue/queue_test.go

```diff
@@ -159,3 +159,72 @@ func TestReinsert(t *testing.T) {
 		t.Errorf("Expected queue to be empty. Has %v items", a)
 	}
 }
+
+func TestFinishProcessingBeforeShutdown(t *testing.T) {
+	q := workqueue.New()
+	q.Add("foo")
+	q.Add("bar")
+
+	firstItem, _ := q.Get()
+	secondItem, _ := q.Get()
+
+	finishedShutdownTest := finishedShutdownTest{}
+	go func() {
+		q.ShutDown()
+		finishedShutdownTest.setFinished()
+	}()
+
+	time.Sleep(20 * time.Millisecond)
+	if !q.ShuttingDown() {
+		t.Fatalf("unexpectedly not indicating shutting down")
+	}
+	if finishedShutdownTest.getFinished() {
+		t.Fatalf("unexpectedly finished ShutDown")
+	}
+
+	// Mark first item done, should not have finished terminating since there's
+	// still one left on the queue
+	q.Done(firstItem)
+	if !q.ShuttingDown() {
+		t.Fatalf("unexpectedly not indicating shutting down")
+	}
+	if finishedShutdownTest.getFinished() {
+		t.Fatalf("unexpectedly finished ShutDown")
+	}
+
+	// Mark second item done, should have finished terminating since there are
+	// no items left on the queue
+	q.Done(secondItem)
+	// Adding a new item on the queue after the shut down is not allowed
+	q.Add("baz")
+	_, shuttingDown := q.Get()
+	if !shuttingDown {
+		t.Fatalf("unexpectedly not indicating shutting down for Get item")
+	}
+	// Sleep a maximum of 1 second since that's the interval used to check the
+	// queue size
+	time.Sleep(1 * time.Second)
```

Comment:
This will be racy. Turn `finishedShutdownTest` into a `sync.WaitGroup`, call `.Wait()` here, time how long it takes if you really want to, but leave at least a 3x margin if you want to fail if it takes too long (I don't think that's necessary).